packages/x-date-pickers/src/hooks/useClearableField.tsx

```diff
@@ -0,0 +1,92 @@
+import * as React from 'react';
+
+import { useSlotProps } from '@mui/base/utils';
+import IconButton from '@mui/material/IconButton';
+import InputAdornment from '@mui/material/InputAdornment';
+import { ClearIcon } from '../icons';
+import {
+  FieldSlotsComponents,
+  FieldSlotsComponentsProps,
+  FieldsTextFieldProps,
+} from '../internals';
+
+type UseClearableFieldProps<
+  TFieldProps extends FieldsTextFieldProps,
+  TInputProps extends { endAdornment?: React.ReactNode } | undefined,
+  TFieldSlots extends FieldSlotsComponents,
+  TFieldSlotsComponentsProps extends FieldSlotsComponentsProps,
+> = {
+  clearable: boolean;
+  fieldProps: TFieldProps;
+  InputProps: TInputProps;
+  onClear: React.MouseEventHandler<HTMLButtonElement>;
+  slots?: { [K in keyof TFieldSlots as Uncapitalize<K & string>]: TFieldSlots[K] };
+  slotProps?: TFieldSlotsComponentsProps;
+};
+
+export const useClearableField = <
+  TFieldProps extends FieldsTextFieldProps,
+  TInputProps extends { endAdornment?: React.ReactNode } | undefined,
+  TFieldSlotsComponents extends FieldSlotsComponents,
+  TFieldSlotsComponentsProps extends FieldSlotsComponentsProps,
+>({
+  clearable,
+  fieldProps: forwardedFieldProps,
+  InputProps: ForwardedInputProps,
+  onClear,
+  slots,
+  slotProps,
+}: UseClearableFieldProps<
+  TFieldProps,
+  TInputProps,
+  TFieldSlotsComponents,
+  TFieldSlotsComponentsProps
+>) => {
+  const EndClearIcon = slots?.clearIcon ?? ClearIcon;
+  const endClearIconProps = useSlotProps({
+    elementType: ClearIcon,
+    externalSlotProps: slotProps?.clearIcon,
+    ownerState: {},
+  });
+
+  const InputProps = {
+    ...ForwardedInputProps,
+    endAdornment: clearable ? (
+      <React.Fragment>
+        <InputAdornment
+          position="end"
+          sx={{ marginRight: ForwardedInputProps?.endAdornment ? -1 : -1.5 }}
```

Comment:
Seems like a nice compromise. 🤔 
Although, we'll probably eventually get a request for an option to provide additional props here, but IMHO, it should be a pretty rare case. 🙈

Reply:
It does feel like a compromise 🤔 Not sure if it is a good one, but regarding the `InputAdornment` spacing issue, this seems like the smallest compromise I think. Do you have a better idea?

Reply:
I'm not sure I can think of any better compromise at the moment. 🤷

Reply:
On the other hand, this solution seems to be a bit intrusive. 🙈 

Do you have ideas on how to solve this inconsistent spacing problem? 🤷 🤔 

https://github.com/mui/mui-x/assets/4941090/b63afb27-3b02-4a47-b0b1-e1dd0c2ad4fb

Reply:
If i understand correctly that margin needs to be adjusted according to the size of the icon when present. It might be possible to extract that size from the icon ?

Reply:
@noraleonte What do you think about the following change?
```diff
diff --git a/docs/data/date-pickers/custom-field/PickerWithBrowserField.tsx b/docs/data/date-pickers/custom-field/PickerWithBrowserField.tsx
index d79ec38ec..adbcc1743 100644
--- a/docs/data/date-pickers/custom-field/PickerWithBrowserField.tsx
+++ b/docs/data/date-pickers/custom-field/PickerWithBrowserField.tsx
@@ -146,7 +146,7 @@ const BrowserSingleInputDateRangeField = React.forwardRef(
         InputProps: {
           ...fieldProps.InputProps,
           endAdornment: (
-            <IconButton onClick={onAdornmentClick}>
+            <IconButton onClick={onAdornmentClick} sx={{ marginLeft: 1 }}>
               <DateRangeIcon />
             </IconButton>
           ),
@@ -227,13 +227,13 @@ const BrowserMultiInputDateRangeField = React.forwardRef(
     } = props;
 
     const { inputRef: startInputRef, ...startTextFieldProps } = useSlotProps({
-      elementType: null as any,
+      elementType: 'input',
       externalSlotProps: slotProps?.textField,
       ownerState: { ...props, position: 'start' },
     }) as MultiInputFieldSlotTextFieldProps;
 
     const { inputRef: endInputRef, ...endTextFieldProps } = useSlotProps({
-      elementType: null as any,
+      elementType: 'input',
       externalSlotProps: slotProps?.textField,
       ownerState: { ...props, position: 'end' },
     }) as MultiInputFieldSlotTextFieldProps;
diff --git a/docs/data/date-pickers/custom-field/PickerWithJoyField.tsx b/docs/data/date-pickers/custom-field/PickerWithJoyField.tsx
index e4f9070dc..dcd76b62d 100644
--- a/docs/data/date-pickers/custom-field/PickerWithJoyField.tsx
+++ b/docs/data/date-pickers/custom-field/PickerWithJoyField.tsx
@@ -134,6 +134,7 @@ type JoySingleInputDateRangeFieldComponent = ((
 const JoySingleInputDateRangeField = React.forwardRef(
   (props: JoySingleInputDateRangeFieldProps, ref: React.Ref<HTMLInputElement>) => {
     const { slots, slotProps, onAdornmentClick, ...other } = props;
+    const { size, color, sx, variant, ...ownerState } = props;
 
     const {
       inputRef: externalInputRef,
@@ -142,10 +143,10 @@ const JoySingleInputDateRangeField = React.forwardRef(
       Dayjs,
       JoyFieldProps & { inputRef: React.Ref<HTMLInputElement> }
     > = useSlotProps({
-      elementType: null as any,
+      elementType: 'input',
       externalSlotProps: slotProps?.textField,
       externalForwardedProps: other,
-      ownerState: props as any,
+      ownerState,
     });
 
     const {
@@ -179,7 +180,12 @@ const JoySingleInputDateRangeField = React.forwardRef(
       <JoyField
         {...processedFieldProps}
         endDecorator={
-          <IconButton onClick={onAdornmentClick} variant="plain" color="neutral">
+          <IconButton
+            onClick={onAdornmentClick}
+            variant="plain"
+            color="neutral"
+            sx={{ marginLeft: 1.5 }}
+          >
             <DateRangeIcon color="action" />
           </IconButton>
         }
@@ -279,13 +285,13 @@ const JoyMultiInputDateRangeField = React.forwardRef(
     } = props;
 
     const { inputRef: startInputRef, ...startTextFieldProps } = useSlotProps({
-      elementType: null as any,
+      elementType: FormControl,
       externalSlotProps: slotProps?.textField,
       ownerState: { ...props, position: 'start' },
     }) as MultiInputFieldSlotTextFieldProps;
 
     const { inputRef: endInputRef, ...endTextFieldProps } = useSlotProps({
-      elementType: null as any,
+      elementType: FormControl,
       externalSlotProps: slotProps?.textField,
       ownerState: { ...props, position: 'end' },
     }) as MultiInputFieldSlotTextFieldProps;

```

Reply:
So we would basically be counteracting the original negative margins? 😬 I don't love it, but I honestly don't see any other way 🤔

Reply:
In the case of Joy UI, we'd just add the necessary styles, because it doesn't have any wrappers with negative margins.